### PR TITLE
metrics: Add nuttcp bandwidth tests

### LINKED ISSUE #540: metrics: Add nuttcp bandwidth tests
Add network bandwidth tests for different buffer sizes with nuttcp performance tool

### PR DESCRIPTION
**Pull request by GabyCT**
Add network bandwidth tests for different buffer sizes with
nuttcp performance tool.

Fixes #540

Signed-off-by: Gabriela Cervantes <gabriela.cervantes.tellez@intel.com>

### CODE REVIEW

File: metrics/network/network-metrics-nuttcp.sh [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+function save_config {
+	metrics_json_start_array
+
+	local json="$(cat << EOF
+	{
+		"image": "$image",
+		"transmit timeout": "$transmit_timeout",
+		"nuttcp version": "7.3.2"
```
Comment (jodh-intel) [inline metrics/network/network-metrics-nuttcp.sh]: There are *three* instances of this version string in this PR. Please could you find a way to define it once?

**Comment by GabyCT** [comment]
@jodh-intel changes applied

**Comment by jodh-intel** [comment]
Thanks @GabyCT.

I just noticed that this PR doesn't actually call the new `metrics/network/network-metrics-nuttcp.sh` test. Do you need to update https://github.com/kata-containers/tests/blob/master/.ci/run_metrics_PR_ci.sh or is this test purely "standalone" / to "run by hand"?

If it's the latter, I think we need some documentation to explain how this test is supposed to be used, particularly after @grahamwhaley's "call to arms" yesterday to try to get more folk involved in metrics.

I guess you could either update https://github.com/kata-containers/tests/blob/master/metrics/README.md, or potentially create a new https://github.com/kata-containers/tests/blob/master/metrics/network/README.md (linked from https://github.com/kata-containers/tests/blob/master/metrics/README.md).

**Comment by GabyCT** [comment]
@jodh-intel, this test will not be part of the CI that is why I am not including in  https://github.com/kata-containers/tests/blob/master/.ci/run_metrics_PR_ci.sh , however, we need these tests to have a reference in network.  I am currently working in the proper documentation and this will be another PR (https://github.com/kata-containers/tests/issues/529)

**Comment by jodh-intel** [comment]
@GabyCT - ack. Going forward, we are really going to have to ensure we don't land tests without docs though.

**Comment by GabyCT** [comment]
@jodh-intel , I already submitted the doc https://github.com/kata-containers/tests/pull/549

**Comment by jodh-intel** [comment]
Hi @GabyCT - great. Personally, I'd combine the two and pull the doc into this PR since that doc relates to the new test you added here.

**Comment by GabyCT** [comment]
@jodh-intel done :)

File: metrics/network/README.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# Kata Containers Iperf3 and Nuttcp network metrics
+
+Currently, Kata Containers have a series of network performance tests. Use
```
Comment (jodh-intel) [inline metrics/network/README.md]: I'd simplify to something like:

> Kata Containers provides a series of network performance tests. Running these provides a basic reference for measuring network essentials like bandwidth, jitter, packet per second throughput and latency.

File: metrics/network/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+bandwidth test shows the speed of the data transfer. The jitter test measures the
+variation in the delay of received packets. The packet-per-second tests show the
+maximum number of (smallest sized) packets allowed through the transports.
+
+- `network-metrics-nuttcp.sh` measures the UDP bandwidth using nuttcp. This tool
+shows the speed of the data transfer for the UDP protocol.
+ 
+## Running the tests
+
+Individual tests run by hand, for example:
```
Comment (jodh-intel) [inline metrics/network/README.md]: s/tests run by hand/tests can be run by hand/

File: metrics/network/README.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+```
+$ cd metrics
+$ bash network/network-metrics-nuttp.sh
+
```
Comment (jodh-intel) [inline metrics/network/README.md]: Nit: extraneous blank line.

File: metrics/network/network-metrics-nuttcp.sh [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+	metrics_json_init
+	save_config
+
+	metrics_json_start_array
+
+	local result_line=$(echo "$result" | tail -1)
+	local -a results
+	read -a results <<< ${result_line%$'\r'}
+
+	# 6, 7, 16, "%"
```
Comment (jodh-intel) [inline metrics/network/network-metrics-nuttcp.sh]: I'd remove this comment.

File: metrics/network/network-metrics-nuttcp.sh [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+		h)
+			help
+			exit 0;
+			;;
+		\?)
+			echo "An invalid option has been entered: -$OPTARG";
+			help
+			exit 1;
+			;;
+		:)
```
Comment (jodh-intel) [inline metrics/network/network-metrics-nuttcp.sh]: :)

File: metrics/network/nuttcp_dockerfile/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+# Usage: FROM [image name]
+FROM ubuntu
+
+# Version of the Dockerfile
+LABEL DOCKERFILE_VERSION="1.0"
+
+# Version of nuttcp
+ARG NUTTCP_VERSION="7.3.2.c"
```
Comment (jodh-intel) [inline metrics/network/nuttcp_dockerfile/Dockerfile]: Ah - I thought this looked a bit odd :) Please can you make this:

```
ARG NUTTCP_VERSION="7.3.2"
  :

gcc nuttcp-${NUTTCP_VERSION}.c -o nuttcp
```

I still worry that we have two code locations with the same values - it's going to be difficult to ensure they stay in sync I think.

**Comment by jodh-intel** [comment]
@klynnrif - ptal.

File: metrics/network/README.md (hunk 2 above)
Comment (klynnrif) [reply metrics/network/README.md]: I agree with the rewrite suggested by @jodh-intel from a grammar rewrite perspective + an additional series comma after "throughput" :)

File: metrics/network/README.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+```
+
+## Expected results
+
+In order to obtain repeteable and stable results is necessary to run the
```
Comment (klynnrif) [inline metrics/network/README.md]: misspelling: repeteable => repeatable and add "it": 
In order to obtain repeatable and stable results it is necessary to run the...

File: metrics/network/network-metrics-nuttcp.sh (hunk 6 above)
Comment (GabyCT) [reply metrics/network/network-metrics-nuttcp.sh]: @jodh-intel well this is funny but in my editor that is the same like you are proposing

**Comment by GabyCT** [comment]
@klynnrif and @jodh-intel changes applied

File: metrics/network/network-metrics-nuttcp.sh (hunk 6 above)
Comment (grahamwhaley) [reply metrics/network/network-metrics-nuttcp.sh]: I think (but we will find out in the morning....) that @jodh-intel may just have been smiling at the smiley :smiley:

Comment (egernst) [reply metrics/network/network-metrics-nuttcp.sh]: Can we add a -v option to run the test with extra detail?  Or, is it possible to output what is actually being run on the client/server?

I'm okay with a follow on PR but wanted to start the conversation...

File: metrics/network/network-metrics-nuttcp.sh [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+	local server_extra_args="--name=$server_name"
+	local client_extra_args="--rm"
+
+	local server_command="tail -f /dev/null"
+	local server_address=$(start_server "$image" "$server_command" "$server_extra_args")
+
+	local client_command="/root/nuttcp -T${transmit_timeout} -u -Ru -i1 -l${bl} ${server_address}"
+	local server_command="/root/nuttcp -u -S"
+
+	docker exec ${server_name} sh -c "${server_command}"
```
Comment (egernst) [inline metrics/network/network-metrics-nuttcp.sh]: As a user, it'd be helpful for me to see what actual nuttcp arguments are used in the calls.  It'd help for quickly rerunning by hand, or slightly modifying but keeping apples to apples....

File: metrics/network/network-metrics-nuttcp.sh (hunk 6 above)
Comment (grahamwhaley) [reply metrics/network/network-metrics-nuttcp.sh]: that's not a terrible idea :-) So far I've tried to keep the metrics scripts fairly 'quiet' by default so we don't pollute the CI logs too much, and when I run them by hand I tend to use a `bash -x test.sh` to get a verbose output. That works fine, apart from we don't get `-x` propagated into any sub-scripts.

Comment (jodh-intel) [reply metrics/network/network-metrics-nuttcp.sh]: Yep - sorry @GabyCT, as @grahamwhaley thought I was just smiling at the smiley! :)

**Pull request merged**